### DISCUSSION TITLE
Welcome to ctrl-M Discussions!

### DISCUSSION THREAD

**Discussion by nuess0r** [nuess0r/ctrl-M#5] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with users of the ctrl-M controller.
  Feels free to:
  * Ask questions you’re wondering about before you buy/build a ctrl-M
  * Ask questions about the Model M modification process that are not covered/unclear in the assembly instructions
  * Share ideas
  * Share your QMK keymaps
  * Engage with other community members.

  As a general note:
  Welcome others and be open-minded. Remember that this is a community we  build together 💪.

**Comment by nuess0r** [comment]
Sorry for the inconvenience.

The missing \ has an easy explanation, I shipped you the ctrl-M
configured for an ISO keyboard like we use it here in Switzerland.
The QMK firmware for the ctrl-M includes support for ANSI keyboards
as well. Unfortunately my QMK firmware is not yet in QMK mainline and
therefore you can't use the QMK online configurator or VIA to easily
fix it.

I updated the QMK firmware for the ctrl-M so by default it supports now
all keys of US (ANSI) and non-US (ISO) variants.
Future ctrl-M shipments will not have this issue.

You can try to update your already assembled ctrl-M (doesn't need to be
installed in the keyboard):

This are the steps needed on a Linux host (I've no experience with
building QMK on Windows):

git clone https://github.com/nuess0r/qmk_firmware.git
git clone https://github.com/nuess0r/qmk_userspace.git

cd qmk_firmware
git checkout ctrl_m
python -m venv .
source bin/activate
pip install qmk
qmk setup
qmk config user.overlay_dir="$(realpath ../qmk_userspace)"
qmk userspace-add -kb ibm/model_m/ctrl_m -km nuess0r
qmk flash -kb ibm/model_m/ctrl_m -km nuess0r

Unplug the ctrl-M USB cable, hold the Esc key and plug the cable in
again.

--> now it should start to download the new firmware should to the
keyboard.  

Download the updated VIA config file:
https://raw.githubusercontent.com/nuess0r/keyboards/refs/heads/master/src/ibm/model_m/ctrl_m/ctrl_m.json

And configure your Model M in Chrome(ium)
https://usevia.app/design

Am Tue, 29 Oct 2024 07:41:00 -0700
schrieb Senator-Asshat ***@***.***>:

> I successfully assembled the ctrl-M board, and installed it in my
> Model M manufactured in May 1987.  All the keys work except for the \
> key.  I moved the keyboard to another computer, and the same issue
> occurs. Also, I tested the ctrl-M with Windows 11 and Ubuntu 24.04
> Linux, and the \ key still does not work. I tried using multiple USB
> cables on different computers and I still cannot use the \ key.
> 
> I followed these instructions
> (https://brain4free.org/wiki/doku.php/elektronik:ctrl_m#discussion__section) to enter the bootloader.  I unplugged the USB cable, and re-plugged it and the issue persists.
> 
> I removed the ctrl-M board, and re-installed the original 139993
> controller and the \ works.  This indicates to me the keyboard
> membrane does not have any issues.
> 
> Do you have any suggestions as to how this issue can be corrected?
> 
> Thank you for any help you can provide.
>

**Reply by Senator-Asshat** [reply]
Thank you very much for this information!  Before I had a chance to ready your response, I replaced the GitHub qmk_firmware with your latest qmk_firmware (https://github.com/nuess0r/qmk_firmware/tree/ctrl_m).  I compiled your updated code into new firmware. Finally, I used a Windows computer and QMK Toolbox to flash the firmware.  Thank you very much for creating this controller! My IBM Model M keyboard will probably outlast me. :)

**Comment by acampbellpayne** [comment]
Thanks to the exchange above I was able to fix my first issue by flashing the firmware (\ not registering on my ANSI-format 1391401) - cheers!

I'm running into trouble when I try to use VIA, error log here if that's helpful:
[VIA-app-errors.csv](https://github.com/user-attachments/files/17656158/VIA-app-errors.csv)

Basically, I'm able to grant access permission in Chrome's security dialog and it even appears as "paired", but it's stuck on the "Authorize device +" screen after granting permission. Quitting and reopening Chrome doesn't seem to change anything here.

**Reply by acampbellpayne** [reply]
Update: checked it with the VIA desktop app, no luck there, either.

**Reply by nuess0r** [reply]
Hi acampbellpayne

Have you loaded the JSON file for the ctrl-M into VIA?

Unfortunately the ctrl-M QMK firmware pull request is still not merged. Therefore I'm not even allowed to create the according pull request for VIA. So for the moment you have to load the ctrl-M VIA definition manually. In the usevia.app go to develop and upload this file https://github.com/nuess0r/keyboards/blob/master/src/ibm/model_m/ctrl_m/ctrl_m.json
Then your keyboard should be loaded and you can configure it.

If you want to you could nicely ask here why it takes so long to merge: https://github.com/qmk/qmk_firmware/pull/24375

**Reply by acampbellpayne** [reply]
Hmm so when I load the definition, this is what I'm seeing in VIA:
![Screenshot from 2024-11-07 15-40-43](https://github.com/user-attachments/assets/9a26697b-e7a7-43c2-ba19-a44aca1f68cd)

Basic functionality like toggling ISO Enter and viewing the matrix overlay seem to be working, but I don't seem to be able to access any of the core functionality. When I load the definition, it prompts me to select a device, but when I select the ctrl-M I get the "warning" icon that is visible in the screenshot, along with the same errors as in the .csv

**Reply by nuess0r** [reply]
I guess you are using Linux. This is a common and misleading error message coming from missing access rights. You can fix that by adding a UDEV rule as described here: https://brain4free.org/wiki/doku.php/elektronik:ctrl_m#linux_udev_rule

**Reply by acampbellpayne** [reply]
ah, thanks - i ended up successfully configuring it on my mac machine. i’ll try that udev config later, looks straightforward enough. thanks for all your help!

**Comment by bbphd** [comment]
I see the kit listed for sale online, but "out of stock". Is someone selling the kit or assembled controllers?

**Reply by nuess0r** [reply]
Hi

Thanks for reaching out.

I sell the kit on Tindie: https://www.tindie.com/products/brain4free/ctrl-m/

I can ship it to most countries but since last summer it's impossible for me to send it to the USA.

My friend at clickykeyboards.com was also selling it inside the USA: https://www.clickykeyboards.com/product/ctrl-m-controller-for-model-m-diy-kit-usb/

At the moment he is out of stock. But we are working on it.
I'll update you when the ctrl-M is available again in the USA.
